server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,285 @@
+      long lastBlockId,
+      long readBufferSize,
+      List<BufferSegment> bufferSegments,
+      List<ShufflePartitionedBlock> resultBlocks) {
+    long nextBlockId = lastBlockId;
+    List<Long> sortedEventId = sortFlushingEventId();
+    int offset = 0;
+    boolean hasLastBlockId = false;
+    // read from inFlushBlockMap first to make sure the order of
+    // data read is according to the order of data received
+    if (!inFlushBlockMap.isEmpty()) {
+      for (Long eventId : sortedEventId) {
+        // update bufferSegments with current blocks in flush map
+        CachedBlocksReadInfo cachedBlocksReadInfo = updateBufferSegmentsWithCachedBlocks(
+            offset, inFlushBlockMap.get(eventId), nextBlockId, readBufferSize,
+            bufferSegments, resultBlocks);
+        offset = cachedBlocksReadInfo.getOffsetInResultData();
+        // if has data or find last blockId, read from begin with next cached blocks
+        if (offset > 0 || cachedBlocksReadInfo.isHasLastBlockId()) {
+          // read from begin in next cached blocks
+          nextBlockId = Constants.INVALID_BLOCK_ID;
+          hasLastBlockId = true;
+        }
+        if (offset >= readBufferSize) {
+          break;
+        }
+      }
+    }
+    // try to read from cached blocks which is not in flush queue
+    if (blocks.size() > 0 && offset < readBufferSize) {
+      CachedBlocksReadInfo cachedBlocksReadInfo = updateBufferSegmentsWithCachedBlocks(
+          offset, blocks, nextBlockId, readBufferSize, bufferSegments, resultBlocks);
+      offset = cachedBlocksReadInfo.getOffsetInResultData();
+      hasLastBlockId = cachedBlocksReadInfo.isHasLastBlockId();
+    }
+    if (lastBlockId != Constants.INVALID_BLOCK_ID && offset == 0 && !hasLastBlockId) {
+      // can't find lastBlockId, it should be flushed
+      // try read again with blockId = Constants.INVALID_BLOCK_ID
+      updateBufferSegmentsAndResultBlocks(
+          Constants.INVALID_BLOCK_ID, readBufferSize, bufferSegments, resultBlocks);
+    }
+  }
+
+  private void updateShuffleData(List<ShufflePartitionedBlock> readBlocks, byte[] data) {
+    int offset = 0;
+    for (ShufflePartitionedBlock block : readBlocks) {
+      // fill shuffle data
+      try {
+        System.arraycopy(block.getData(), 0, data, offset, block.getLength());
+      } catch (Exception e) {
+        LOG.error("Unexpect exception for System.arraycopy, length["
+            + block.getLength() + "], offset["
+            + offset + "], dataLength[" + data.length + "]", e);
+        throw e;
+      }
+      offset += block.getLength();
+    }
+  }
+
+  private List<Long> sortFlushingEventId() {
+    List<Long> eventIdList = Lists.newArrayList(inFlushBlockMap.keySet());
+    eventIdList.sort((id1, id2) -> {
+      if (id1 > id2) {
+        return 1;
+      }
+      return -1;
+    });
+    return eventIdList;
+  }
+
+  private CachedBlocksReadInfo updateBufferSegmentsWithCachedBlocks(
```

Comment:
better name？

Reply:
`updateBufferSegmentsAndResultBlocks`?

Reply:
OK, CacheBlocksReadInfo should be a paramter?

Reply:
The current name is `readCachedBlocksAndUpdateSegments`

Reply:
separate two methods.

Reply:
ok